compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+//     //   if (Object.keys(fnTy.args).length !== args.length) {
+//     //     throw new Error('~~ Minecraft Villager sad noise :( ~~');
+//     //   }
+//     //   for (let ii = 0; ii < args.length; ii++) {
+//     //     let argName = args[ii].dec.name;
+//     //     let argMeta = metadata.var(argName);
+//     //     if (argMeta.dec !== args[ii].dec) {
+//     //       throw new Error('invalid call state: arg ref and var def mismatch');
+//     //     }
+//     //     let paramTy = Object.values(fnTy.args)[ii].ty;
+//     //     argMeta.constraints.push(paramTy);
+//     //   }
+//     // }
+//     stmts.push(call);
+
+//     return stmts;
+//   }
+
+//   inline(amm: Output, assignName?: string, assignTy?: Type) {
+//     TODO('inline fns?');
+//   }
+
+//   exprTy(): Type {
+//     return this.retTy;
+//   }
+// }
+
+export class Closure extends Stmt {
+  exprTy(): Type {
+    return TODO('closures')
+  }
+
+  inline(amm: Output) {
+    TODO('closure inlining');
+  }
+}
+
+export class Cond extends Stmt {
+  static fromConditionalsAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    return TODO('build conditionals');
+  }
+
+  exprTy(): Type {
+    throw new Error(`conditionals can't be used as expressions`);
+  }
+
+  inline(amm: Output) {
+    TODO('conditional inlining');
+  }
+}
+
+export class Dec extends Stmt {
+  private __ammName: string
+  mutable: boolean
+  name: string
+  ty: Type
+  val: Stmt
+
+  get ammName(): string {
+    return this.__ammName;
+  }
+
+  set ammName(to: string) {
+    this.__ammName = to;
+  }
+
+  constructor(
+    ast: LPNode,
+    mutable: boolean,
+    name: string,
+    ty: Type | null = null,
+    val: Stmt,
+  ) {
+    super(ast);
+    this.mutable = mutable;
+    this.name = name;
+    this.ty = ty !== null ? ty : Type.generate();
+    this.val = val;
+    this.ammName = genName();
+  }
+
+  static fromAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    let work: LPNode = null;
+    let mutable: boolean;
+    if (ast.has('constdeclaration')) {
+      work = ast.get('constdeclaration');
+      mutable = false;
+    } else {
+      work = ast.get('letdeclaration');
+      mutable = true;
+    }
+    const name = work.get('variable').t;
+    let ty: Type = null;
+    if (work.has('typedec')) {
+      const tyName = work.get('typedec').get('fulltypename');
+      ty = Type.getFromTypename(tyName, metadata.scope);
+    }
+    stmts.push(...Stmt.fromAssignables(work.get('assignables'), metadata));
+    let dec = stmts.pop();
+    if (dec instanceof Dec) {
+      metadata.variables[name] = metadata.var(dec.name);
+      metadata.variables[dec.name] = null;
+      dec.ty.constrain(ty);
+      dec.mutable = mutable;
+      dec.name = name;
+    } else if (dec instanceof Ref) {
+      const upTy = dec.ty;
+      dec = new Dec(
+        ast,
+        mutable,
+        name,
+        ty,
+        dec,
+      );
+      (dec as Dec).ty.constrain(upTy);
```

Comment:
Is this explicit type override actually necessary here? You literally declare a `new Dec` right before this?

Reply:
you would think? but no, unfortunately TS was giving me a type error :(